engine/jobmaster/dm/metadata/ddl.go

```diff
@@ -14,35 +14,195 @@
 package metadata
 
 import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/pingcap/tiflow/engine/pkg/adapter"
 	metaModel "github.com/pingcap/tiflow/engine/pkg/meta/model"
+	"github.com/pingcap/tiflow/pkg/errors"
 )
 
-// DDL represents the state of ddls.
-// TODO: implement DDL
-type DDL struct{}
+// SourceTable represents a upstream table
+type SourceTable struct {
+	Source string
+	Schema string
+	Table  string
+}
+
+// MarshalText implements json.MarshalText.
+func (st SourceTable) MarshalText() (text []byte, err error) {
+	type s SourceTable
+	return json.Marshal(s(st))
```

Comment:
Can we use

```suggestion
	return json.Marshal(st)
```

or remove this function to use default marshaller?

Reply:
No, we want to use SourceTable as key https://github.com/pingcap/tiflow/blob/7521c5107d8a5f551520a9b7d2afcbca0ded9606/engine/jobmaster/dm/metadata/ddl.go#L85
From [json doc](https://pkg.go.dev/encoding/json#Marshal), `Map values encode as JSON objects. The map's key type must either be a string, an integer type, or implement encoding.TextMarshaler.`